src/content/app/tools/blast/state/epics/blastEpics.ts

```diff
@@ -0,0 +1,184 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  from,
+  timer,
+  Subject,
+  pipe,
+  map,
+  mergeMap,
+  concatMap,
+  filter,
+  expand,
+  tap,
+  toArray,
+  EMPTY
+} from 'rxjs';
+import { isFulfilled, type Action } from '@reduxjs/toolkit';
+import type { Epic } from 'redux-observable';
+
+import config from 'config';
+
+import { submitBlast } from '../blast-api/blastApiSlice';
+
+import * as observableApiService from 'src/services/observable-api-service';
+import {
+  saveBlastSubmission,
+  updateSavedBlastJob
+} from 'src/content/app/tools/blast/services/blastStorageService';
+
+import {
+  updateJob,
+  restoreBlastSubmissions,
+  type JobStatus,
+  type BlastJob
+} from 'src/content/app/tools/blast/state/blast-results/blastResultsSlice';
+
+import type { RootState } from 'src/store';
+
+const POLLING_INTERVAL = 15000; // fifteen seconds
+
+/**
+ * An epic that listens to redux action with the data from a BLAST form submission.
+ * It saves the submission to IndexedDB, polls for job statuses,
+ * and when a job is finished or failed, updates its status in indexedDB,
+ * and sends a redux action to update its status in redux
+ */
+export const blastFormSubmissionEpic: Epic<Action, Action, RootState> = (
+  action$
+) =>
+  action$.pipe(
+    filter(submitBlast.matchFulfilled),
+    tap((action) => {
+      // save BLAST submission to IndexedDB
+      const { submissionId, submission } = action.payload;
+      saveBlastSubmission(submissionId, submission);
+    }),
+    map((action) => {
+      const {
+        submissionId,
+        submission: { results }
+      } = action.payload;
+
+      return results.map((job) => ({ submissionId, job }));
+    }),
+    poll(),
+    tap(databaseUpdaterSubject),
+    map((pollingResult) => {
+      const {
+        submissionId,
+        job: { jobId, status }
+      } = pollingResult;
+      // finish by returning a redux action
+      return updateJob({ submissionId, jobId, fragment: { status } });
+    })
+  );
+
+/**
+ * Similar to blastFormSubmissionEpic, but listens to redux action
+ * that reads BLAST submissions back from IndexedDB.
+ * It checks for running jobs among the saved submissions,
+ * and polls their status until their are finished or failed.
+ */
+export const blastSubmissionsResroteEpic: Epic<Action, Action, RootState> = (
+  action$
+) =>
+  action$.pipe(
+    filter(isFulfilled(restoreBlastSubmissions)),
+    map(({ payload: submissions }) => {
+      return Object.entries(submissions).flatMap(([submissionId, submission]) =>
+        submission.results
+          .filter((job) => job.status === 'RUNNING')
+          .map((job) => ({
+            submissionId,
+            job
+          }))
+      );
+    }),
+    poll(),
+    tap(databaseUpdaterSubject),
+    map((pollingResult) => {
```

Comment:
> would it be a bad idea to call updateJob within databaseUpdaterSubject to avoid the map on L114?

A redux-observable epic is a stream that starts with some redux action(s) and ends with another redux action. In our case, the final action that comes out of the epic is the return value of`updateJob`.

Updating IndexedDB is a side effect, which is why I put it in a `tap` function. `tap` is an identity function that returns the same value as it receives; but it allows you to perform whatever logic you want inside. Sadly, it's not asynchronous, whereas updating IndexedDB is an asynchronous operation; which is why I created a dedicated subject that can control the rate of indexeddb updates such that the next one will happen only after the first.